File: content-solutions/graph-advanced/mst-manhattan.py
def solution(points):
    n = len(points)
    parent = list(range(n))
    size = [1] * n

    def find(x):
        root = x
        while parent[root] != root:
            root = parent[root]
        while x != root:
            nxt = parent[x]
            parent[x] = root
            x = nxt
        return root

    def union(a, b):
        ra, rb = find(a), find(b)
        if ra == rb:
            return False
        if size[ra] < size[rb]:
            ra, rb = rb, ra
        parent[rb] = ra
        size[ra] += size[rb]
        return True

    edges = []
    for i in range(n):
        x1, y1 = points[i]
        for j in range(i + 1, n):
            x2, y2 = points[j]
            edges.append((abs(x1 - x2) + abs(y1 - y2), i, j))
    edges.sort()
    total = 0
    picked = 0
    for d, i, j in edges:
        if union(i, j):
            total += d
            picked += 1
            if picked == n - 1:
                break
    return total

File: content-solutions/graph-advanced/mst-manhattan.java
import java.util.*;

class Solution {
    int[] parent, size;

    void init(int n) {
        parent = new int[n];
        size = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            size[i] = 1;
        }
    }

    int find(int x) {
        int root = x;
        while (parent[root] != root) root = parent[root];
        while (x != root) {
            int next = parent[x];
            parent[x] = root;
            x = next;
        }
        return root;
    }

    boolean union(int a, int b) {
        int ra = find(a), rb = find(b);
        if (ra == rb) return false;
        if (size[ra] < size[rb]) {
            int t = ra;
            ra = rb;
            rb = t;
        }
        parent[rb] = ra;
        size[ra] += size[rb];
        return true;
    }

    public long solution(int[][] points) {
        int n = points.length;
        init(n);
        int[][] edges = new int[n * (n - 1) / 2][];
        int k = 0;
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                int d = Math.abs(points[i][0] - points[j][0]) + Math.abs(points[i][1] - points[j][1]);
                edges[k++] = new int[] {d, i, j};
            }
        }
        Arrays.sort(edges, (x, y) -> Integer.compare(x[0], y[0]));
        long total = 0;
        int picked = 0;
        for (int[] e : edges) {
            if (picked == n - 1) break;
            if (union(e[1], e[2])) {
                total += e[0];
                picked++;
            }
        }
        return total;
    }
}
